src/py21cmfast/src/hmf.c

```diff
@@ -856,7 +858,9 @@ double Nion_General(double z, double lnM_Min, double lnM_Max, double MassTurnove
         .HMF = matter_options_global->HMF,
         .gamma_type = 3,
     };
-    return IntegratedNdM(lnM_Min, lnM_Max, params, &u_nion_integrand, 0);
+    // z factor can be taken out of the integral
+    double Zesc = pow((1 + z) / 8.0, sc->beta_esc);
+    return Zesc * IntegratedNdM(lnM_Min, lnM_Max, params, &u_nion_integrand, 0);
```

Comment:
See above comment, redshift-based escape fractions should be pre-computed once per snapshot